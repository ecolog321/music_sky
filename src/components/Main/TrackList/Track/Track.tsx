import { FC } from "react";
import Image from "next/image";
import styles from "./track.module.css";
import { TrackType } from "../../../../types/types";
import { secondsToMs } from "../../../../utils/secondsToMs";
import {
  setCurrentTrack,
  setIsPlaying,
} from "../../../../store/features/playlistSlise";
import { useAppDispatch, useAppSelector } from "../../../../hooks/store";
import clsx from "clsx";
import { useLikeTracks } from "../../../../hooks/useLikeTrack";

type Props = {
  tracks: TrackType[];
  track: TrackType;
};

export const Track: FC<Props> = ({ track, tracks }) => {
  const { name, author, album, duration_in_seconds } = track;
  const dispatch = useAppDispatch();
  const currentTrack = useAppSelector((state) => state.playlist.currentTrack);
  const isPlaying = useAppSelector((state) => state.playlist.isPlaying);
  const isCurrentTrack = currentTrack?._id === track._id;
  const { isLiked, handleLike } = useLikeTracks(track._id);

  const toogleTrack = () => {
    dispatch(setCurrentTrack({ currentTrack: track, currentPlaylist: tracks}));
    dispatch(setIsPlaying(true));
  };

  return (
    <div className={styles.playlist__item}>
      <div className={styles.playlist__track} onClick={toogleTrack}>
        <div className={clsx(styles.track__title, styles.col01)}>
          <div className={styles.track__title_image}>
            <Image
              className={isCurrentTrack ? clsx(styles.dote, isPlaying && styles.dote_playing) : styles.track__title_svg}
              src={isCurrentTrack ? "/img/icons/track.png" : '/img/icons/note.svg'}
              width={24}
              height={24}
              alt="note"
            />
          </div>
          <div className={styles.track__title_text}>
            <div className={styles.track__title_link}>
              <a>{name}</a>
              <span className={styles.track__title_span}></span>
            </div>
          </div>
        </div>
        <div className={clsx(styles.track__author, styles.col02)}>
          <a className={styles.track__author_link}>{author}</a>
        </div>
        <div className={clsx(styles.track__album, styles.col03)}>
          <a className={styles.track__album_link}>{album}</a>
        </div>
        <div className={clsx(styles.track__time, styles.col04)} onClick={handleLike}>
          <svg
            className={clsx(
              isLiked ? styles.track__liked_svg : styles.track__time_svg
            )}
          >
            <use xlinkHref="/img/icons/sprite.svg#icon-like"></use>
          </svg>
          <span className={styles.track__time_text}>
            {secondsToMs(duration_in_seconds)}
          </span>
        </div>
      </div>
    </div>
  );
};
