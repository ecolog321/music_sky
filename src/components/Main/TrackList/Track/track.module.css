.playlist__item {
    width: 100%;
    display: block;
    margin-bottom: 12px;
    cursor: pointer;
  }
  
  .playlist__track {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-pack: justify;
    -ms-flex-pack: justify;
    justify-content: space-between;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }

  .playlist-title__col {
    font-style: normal;
    font-weight: 400;
    font-size: 14px;
    line-height: 24px;
    letter-spacing: 2px;
    color: #696969;
    text-transform: uppercase;
  }
  
  .playlist-title__svg {
    width: 12px;
    height: 12px;
    fill: transparent;
    stroke: #696969;
  }

  .track__title {
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    width: 350px;
  }
  
  .track__title_image {
    width: 51px;
    height: 51px;
    padding: 16px;
    background: #313131;
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: center;
    -ms-flex-pack: center;
    justify-content: center;
    margin-right: 17px;
  }
  
  .track__title_svg {
    width: 18px;
    height: 17px;
    fill: transparent;
    stroke: #4e4e4e;
  }
  
  .track__title_link {
    font-style: normal;
    font-weight: 400;
    font-size: 16px;
    line-height: 24px;
    color: #ffffff;
  }
  
  .track__title_span {
    font-style: normal;
    font-weight: 400;
    font-size: 16px;
    line-height: 24px;
    color: #4e4e4e;
  }
  
  .track__author {
    width: 321px;
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: start;
    -ms-flex-pack: start;
    justify-content: flex-start;
  }
  
  .track__author-link {
    font-style: normal;
    font-weight: 400;
    font-size: 16px;
    line-height: 24px;
    color: #ffffff;
    text-align: left;
  }
  
  .track__album {
    width: 245px;
  }
  
  .track__album_link {
    font-style: normal;
    font-weight: 400;
    font-size: 16px;
    line-height: 24px;
    color: #696969;
  }
  
  .track__time_svg {
    width: 14px;
    height: 12px;
    margin-right: 17px;
    fill: transparent;
    stroke: #696969;
  }

  .track__liked_svg {
    width: 14px;
    height: 12px;
    margin-right: 17px;
    fill: #653596;
    stroke: #653596;
  }
  
  .track__time_text {
    font-style: normal;
    font-weight: 400;
    font-size: 16px;
    line-height: 24px;
    text-align: right;
    color: #696969;
  }

  @keyframes glowing {
    0% { background-color: #D9B6FF; box-shadow: 0 0 3px #D9B6FF; }
    50% { background-color: #AD61FF; box-shadow: 0 0 10px #AD61FF; }
    100% { background-color: #653596; box-shadow: 0 0 3px #653596; }
    }
  
  .dote {
    background-color: #AD61FF;
    border-radius: 20px;
    width: 20px;
    height: 20px;
  }

  .dote_playing {
    background-color: #AD61FF;
    border-radius: 20px;
    width: 20px;
    height: 20px;
    animation: glowing 1300ms infinite;
  }
 

