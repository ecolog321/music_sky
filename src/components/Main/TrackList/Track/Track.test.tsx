
import { isEven, Track } from "./Track";
import renderer from "react-test-renderer";
import ReduxProvider from "../../../../store/ReduxProvider";

describe("track", () => {
  it("render track", () => {
    const track = {
      _id: 1,
      name: "Ora",
      author: "melo",
      release_date: "12-123-2",
      genre: ["1"],
      duration_in_seconds: 10,
      album: "qwetty",
      logo: null,
      track_file: "url/123",
      stared_user: [],
    };
    const tree = renderer
      .create(
        <ReduxProvider>
          <Track track={track} tracks={[]} />
        </ReduxProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("must determine even number",()=>{
    expect(isEven(6)).toBe(true)
    expect(isEven(5)).toBe(false)
    expect(isEven(10)).toBe(true)
    expect(isEven(5.5)).toBe(false)
    expect(isEven(0)).toBe(true)
  })
});
