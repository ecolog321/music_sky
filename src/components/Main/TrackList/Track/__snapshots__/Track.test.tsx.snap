// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="playlist__item"
>
  <div
    className="playlist__track"
    onClick={[Function]}
  >
    <div
      className="track__title"
    >
      <div
        className="track__title_image"
      >
        <img
          alt="note"
          className="track__title_svg"
          data-nimg="1"
          decoding="async"
          height={24}
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="/img/icons/note.svg"
          style={
            {
              "color": "transparent",
            }
          }
          width={24}
        />
      </div>
      <div
        className="track__title_text"
      >
        <div
          className="track__title_link"
        >
          <a>
            Ora
          </a>
          <span
            className="track__title_span"
          />
        </div>
      </div>
    </div>
    <div
      className="track__author"
    >
      <a
        className="track__author_link"
      >
        melo
      </a>
    </div>
    <div
      className="track__album"
    >
      <a
        className="track__album_link"
      >
        qwetty
      </a>
    </div>
    <div
      className="track__time"
      onClick={[Function]}
    >
      <svg
        className="track__time_svg"
      >
        <use
          xlinkHref="/img/icons/sprite.svg#icon-like"
        />
      </svg>
      <span
        className="track__time_text"
      >
        00:010
      </span>
    </div>
  </div>
</div>
`;

exports[`track render track 1`] = `
<div
  className="playlist__item"
>
  <div
    className="playlist__track"
    onClick={[Function]}
  >
    <div
      className="track__title"
    >
      <div
        className="track__title_image"
      >
        <img
          alt="note"
          className="track__title_svg"
          data-nimg="1"
          decoding="async"
          height={24}
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="/img/icons/note.svg"
          style={
            {
              "color": "transparent",
            }
          }
          width={24}
        />
      </div>
      <div
        className="track__title_text"
      >
        <div
          className="track__title_link"
        >
          <a>
            Ora
          </a>
          <span
            className="track__title_span"
          />
        </div>
      </div>
    </div>
    <div
      className="track__author"
    >
      <a
        className="track__author_link"
      >
        melo
      </a>
    </div>
    <div
      className="track__album"
    >
      <a
        className="track__album_link"
      >
        qwetty
      </a>
    </div>
    <div
      className="track__time"
      onClick={[Function]}
    >
      <svg
        className="track__time_svg"
      >
        <use
          xlinkHref="/img/icons/sprite.svg#icon-like"
        />
      </svg>
      <span
        className="track__time_text"
      >
        00:010
      </span>
    </div>
  </div>
</div>
`;
